### PR TITLE
restore-state: extend bit size for `RestoreState`

### PR DESCRIPTION
**Pull request by rmsyn**
Extends the type alias bit size for `RestoreState` types.

Adds the following states:

- `restore-state-u64x2`: 128-bits
- `restore-state-u64x4`: 256-bits
- `restore-state-u64x8`: 512-bits
- `restore-state-u64x16`: 1024-bits

These extensions are necessary for platforms like multicore RISC-V SoCs.

For example, StarFive JH7110 SoCs have five HARTs with 136 interrupts each (680 total). Further extensions will be necessary for truly large SoCs like the 64-core SOPHGO SG2042.

### CODE REVIEW

**Comment by rmsyn** [comment]
Ping @Dirbaio can you please review this PR?

File: src/lib.rs [hunk 1]
```diff
@@ -50,31 +50,60 @@ impl<'cs> CriticalSection<'cs> {
...
+    all(feature = "restore-state-u32", feature = "restore-state-u64x2"),
+    all(feature = "restore-state-u32", feature = "restore-state-u64x4"),
+    all(feature = "restore-state-u32", feature = "restore-state-u64x8"),
+    all(feature = "restore-state-u32", feature = "restore-state-u64x16"),
     all(feature = "restore-state-u32", feature = "restore-state-usize"),
     all(feature = "restore-state-u64", feature = "restore-state-usize"),
+    all(feature = "restore-state-u64x2", feature = "restore-state-usize"),
+    all(feature = "restore-state-u64x4", feature = "restore-state-usize"),
+    all(feature = "restore-state-u64x8", feature = "restore-state-usize"),
+    all(feature = "restore-state-u64x16", feature = "restore-state-usize"),
```
Comment (Dirbaio) [inline src/lib.rs:92]: this is getting a bit out of hand, it's no longer easy to tell at a glance whether this is correct. Can you change it to use [this](https://github.com/embassy-rs/embassy/blob/main/embassy-executor/src/lib.rs#L14-L26)?

(i'd just copypaste it. macro is quite small, it's not worth making a separate crate just for that)

**Comment by Dirbaio** [comment]
- If these platforms are 32bit, I wonder if using u32x4, u32x8... would be more performant.
- I find having such big restore states is a bit extreme. We should add them if there's hardware out there that needs this, but I wonder if it's really necessary? Having to save and restore such a big state qould be quite slow. Don't these SoCs have some "disable all interrupts" bit to avoid disabling/restoring
-  all interrupts individually?

File: src/lib.rs (hunk 1 above)
Comment (rmsyn) [reply src/lib.rs:92]: Awesome, I thought the same, I'll definitely use this helper macro.

**Comment by rmsyn** [comment]
> If these platforms are 32bit, I wonder if using u32x4, u32x8... would be more performant

The platform that prompted these changes is the U74(MC) from SiFive, which is a 64-bit platform. The SG2042 is also 64-bit.

> I find having such big restore states is a bit extreme. We should add them if there's hardware out there that needs this, but I wonder if it's really necessary? Having to save and restore such a big state qould be quite slow. Don't these SoCs have some "disable all interrupts" bit to avoid disabling/restoring all interrupts individually?

So, I ended up switching to the `MIE/SIE` register, and storing the state in a `bool`. After more reading it looks like this is actually a global switch. I originally thought more was needed, because the `single-hart` feature in `riscv-rt` made me think `MIE/SIE` registers only apply to core-local interrupts.

The reason I thought the state needs to be so large is to encode the entire bit-space for the `PLIC` `enable` registers. Section [10.5 Interrupt Enables of the U74-MC Core Complex Manual](https://starfivetech.com/uploads/u74_core_complex_manual_21G1.pdf) describes how the `PLIC` handles the peripheral interrupt mapping. Since there are 136 total interrupts, that means to encode the total space:

- 136-bit (17 x 8-bit)
- 144-bit (9 x 16-bit)
- 160-bit (5 x 32-bit)
- 192-bit (3 x 64-bit)

That would be per-core, so for the total system it's `136-bit * num-cores`.

I think the `MIE/SIE` registers will enable/disable interrupts globally, at least according the U74 Core Complex Manual. Then, it will be up to the user to configure routing for interrupts they want to handle with the `PLIC`. That would be separate from the `RestoreState` stuff, and make this PR unnecessary.

Does this sound right to you?

**Comment by rmsyn** [comment]
@MabezDev thanks for showing me that implementation. I had looked at `esp-hal` for inspiration in writing a UART, and missed the `CriticalSection` stuff.

That makes sense to use a static mutex with a spin lock.

Basically puts the non-owning cores to sleep waiting for their chance to lock the mutex, very nice :)

**Comment by rmsyn** [comment]
@Dirbaio I think this PR can be closed now. If someone finds it useful in the future, feel free to crib at will.

**Pull request closed without merging**